harden console handling

brauner opened the pull request:
Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

brauner commented on src/lxc/conf.c:
```diff
@@ -1587,22 +1577,84 @@ static int lxc_setup_ttydir_console(const struct lxc_rootfs *rootfs,
+	/* When we are asked to setup a console we remove any previous
+	 * /dev/console bind-mounts.
+	 */
+	if (console->path && !strcmp(console->path, "none")) {
```
I guess this codepath is something we want to discuss. We can either take any bind-mount for /dev/console absolutely literally and do not move it around in case `lxc.devttydir` is requested or do as I'm doing here, movie it under `lxc.devttydir`.

lifeng68 replied:
@brauner Hi, i find this PR when i deal with the #2092. 
I do not quit understand why we should do this, if the user specify console->path = none, i think we should do not do anything. The man page of lxc
```
lxc.console.path
Specify a path to a device to which the console will be attached. The keyword 'none' will simply disable the console. Note, when specifying 'none' and creating a device node for the console in the container at /dev/console or bind-mounting the hosts's /dev/console into the container at /dev/console the container will have direct access to the hosts's /dev/console. This is dangerous when the container has write access to the device and should thus be used with caution.
```
I wonder how to  create a device node for the console in the container at /dev/console ? mknod in the container?

brauner replied:
> @brauner Hi, i find this PR when i deal with the #2092.
I do not quit understand why we should do this, if the user specify console->path = none, i think we should do not do anything. The man page of lxc

Agreed! Please add this to your PR https://github.com/lxc/lxc/pull/2092 if you haven't already. :)

> I wonder how to create a device node for the console in the container at /dev/console ? mknod in the container?

You mean at container runtime or during container setup?
- runtime from inside the container: in order to create a console device node at runtime from inside the container you need to retain `CAP_MKNOD`
- runtime from outside the container: needs `CAP_MKNOD` on the host and involves some tricks with mount namespaces + bind-mounting
- during container setup: needs `CAP_MKNOD` on the host but there's currently no logic to create an actual `/dev/console` device. I'm also not sure that this makes sense or even is secure since `/dev/console` is not namespaced and would thus be just another instance of the hosts `/dev/console` device. That's why we always do bind-mounts of pty devices.

brauner commented on src/lxc/conf.c:
```diff
@@ -1587,22 +1577,84 @@ static int lxc_setup_ttydir_console(const struct lxc_rootfs *rootfs,
+	/* When we are asked to setup a console we remove any previous
+	 * /dev/console bind-mounts.
+	 */
+	if (console->path && !strcmp(console->path, "none")) {
```
Fyi, this is what our code is currently *trying* to do but not succeeding at because of the `unlink()` `EBUSY`.

brauner commented:
jenkins: test this please

brauner commented:
As a short overview:

1. privileged
(i)
    ```
    lxc.console=none
    lxc.devttydir=lxc
    ```
    --> no `/dev/console` with `/dev/tty<n>` setup under `/dev/lxc`
(ii)
   ```
    lxc.console=none
    lxc.devttydir=
    ```
    --> no `/dev/console` with `/dev/tty<n>` setup directly under `/dev`
(iii)
    ```
    lxc.console=none
    lxc.devttydir=
    lxc.mount.entry = /dev/console dev/console none bind,create=file 0 0
    ```
    --> grab host's `/dev/console` and setup together with `/dev/tty<n>` directly under `/dev`
(iv)
    ```
    lxc.console=none
    lxc.devttydir=lxc
    lxc.mount.entry = /dev/console dev/console none bind,create=file 0 0
    ```
    --> grab host's `/dev/console` and setup (via `MS_MOVE`) together with `/dev/tty<n>` under `/dev/lxc`
(v)
    ```
    lxc.devttydir=lxc
    lxc.mount.entry = /dev/console dev/console none bind,create=file 0 0
    ```
    --> User requests a console but also a bind-mount for the hosts `/dev/console`. Here we unmount the `/dev/console` bind-mount and, create a fresh empty `/dev/console` file and bind-mount a `/dev/pts/<n>` device over it under `/dev/lxc`.
(vi)
    ```
    lxc.devttydir=lxc
    lxc.mount.entry = /dev/console dev/console none bind,create=file 0 0
    ```
    --> same as (v) but directly under `/dev`.

As a short overview:

2. unprivileged
(i)
    ```
    lxc.console=none
    lxc.devttydir=lxc
    ```
    --> no `/dev/console` with `/dev/tty<n>` setup under `/dev/lxc`
(ii)
   ```
    lxc.console=none
    lxc.devttydir=
    ```
    --> no `/dev/console` with `/dev/tty<n>` setup directly under `/dev`
(iii)
    ```
    lxc.console=none
    lxc.devttydir=
    lxc.mount.entry = /dev/console dev/console none bind,create=file 0 0
    ```
    --> bind-mount host's `/dev/console` and setup together with `/dev/tty<n>` directly under `/dev`
(iv)
    ```
    lxc.console=none
    lxc.devttydir=lxc
    lxc.mount.entry = /dev/console dev/console none bind,create=file 0 0
    ```
    --> bind-mount `/dev/console` and setup (via `MS_MOVE`) together with `/dev/tty<n>` under `/dev/lxc`
(v)
    ```
    lxc.devttydir=lxc
    lxc.mount.entry = /dev/console dev/console none bind,create=file 0 0
    ```
    --> User requests a console but also a bind-mount for the hosts `/dev/console`. Here we unmount the `/dev/console` bind-mount and, create a fresh empty `/dev/console` file and bind-mount a `/dev/pts/<n>` device over it under `/dev/lxc`.
(vi)
    ```
    lxc.devttydir=lxc
    lxc.mount.entry = /dev/console dev/console none bind,create=file 0 0
    ```
    --> same as (v) but directly under `/dev`.

brauner commented:
Also, this series of commits makes it possible to use `lxc.devttydir` with unpriviliged containers.

hallyn commented:
(note that this will fail to build between patches 3 and 5.  let'snot rock the boat by fixing that now, but it could bug someone doing a bisect later)

hallyn commented:
You introduce one use of 'mount', which is not using safe_mount, when move mounting /dev/console under the devttydir.  Normally we create that directory (which should avoid the chance of there being a symlink redirect into rootfs), but if it already exists we use what's there.

So I don't think that's safe - user could create a devttydir containing a symlink into rootfs and cause us to mount into host directory.

brauner commented:
So simply use `safe_mount` is what you're saying?

brauner commented:
@hallyn, replaced `mount()` by `safe_mount()` in the one case.

I also added `lxc_unstack_mountpoint()` to clear all mounts atop `/dev/console` if the user requests one.

brauner commented:
`lxc_unstack_mountpoint()`:

```
lxc-start xen1 20170509224159.533 DEBUG    lxc_conf - conf.c:lxc_setup_ttydir_console:1562 - created directory for console and tty devices at /usr/lib/x86_64-linux-gnu/lxc/dev/bla"
lxc-start xen1 20170509224159.533 DEBUG    lxc_conf - conf.c:lxc_setup_ttydir_console:1622 - moved mount "/usr/lib/x86_64-linux-gnu/lxc/dev/console" to "/usr/lib/x86_64-linux-gnu/lxc/dev/bla/console"
lxc-start xen1 20170509224200.214 DEBUG    lxc_conf - conf.c:lxc_setup_ttydir_console:1631 - cleared all (31) mounts from "/usr/lib/x86_64-linux-gnu/lxc/dev/console"
lxc-start xen1 20170509224200.214 DEBUG    lxc_conf - conf.c:lxc_setup_ttydir_console:1668 - console has been setup under "bla/console" and symlinked to "/usr/lib/x86_64-linux-gnu/lxc/dev/console"
```

brauner commented:
Note, the mount is obviously only moved if it points to a valid `/dev/console c 5 1` device.
